### ISSUE TITLE
Asynchronous, but still janking on large (>2k) lists

### ISSUE THREAD

**Issue by krnlde** [brianmhunt/knockout-fast-foreach#35]
As mentioned here knockout/knockout/issues/1954 your fastForEach implementation has recognizable improvements over the original ko-foreach. But still I experience big janks of 1-2 sec when iterating over a list with 5k entries.

Mind the chrome timeline capture here:
![image](https://cloud.githubusercontent.com/assets/1087002/11744317/94cde36e-a00f-11e5-888f-0344acb2c705.png)

The biggest part of the 2876ms frame is the purple "recalculating style". I didn't get into the "animation frame fired" JS part too deep, but it feels like it is not janking.

**Comment by brianmhunt** [comment]
Interesting, thanks @krnide.  I'd really like to fix that – and it's precisely the kind of scenario I'd like FFE to be comfortable with.

Can you give any more info on what's happening in the recalculating part?  Or perhaps share a jsFiddle/CodePen/etc?

**Comment by brianmhunt** [comment]
Incidentally, if you're using a `<li>` tag does `li {display:block; }` help?  Per http://stackoverflow.com/questions/19796057

**Comment by brianmhunt** [comment]
Incidentally, at one point I was adding/removing 100,000 items from lists, without much janking.  I wonder if there's been a regression.

**Comment by krnlde** [comment]
It is exactly your example: http://output.jsbin.com/dakihezega/2
- Open the site in Chrom[e|ium]
- Open its dev tools
- navigate to "Timeline"
- Hit Record
- check "use fastForEach"
- Enter 5000 in the text box
- Hit add items
- Wait for ~5 seconds
- Stop recording
- Analyze

~~Regarding your hint with the `display: block`: I would pretty much like to use the fastForEach in SVG, which has no `display: block`~~ SVG _has_ `display:block`, but does not do any difference.

**Comment by krnlde** [comment]
btw Firefox behaves exactly the same. Same jank time.

**Comment by brianmhunt** [comment]
Oh yes, I see.  Thanks for clarifying. :)

So the good news is that the time is linear i.e. inserting 500 items is 150ms, 5000 items is 1500ms.  The problem, then, is not the insertion of items but the rendering of off-screen items (whose rendering needn't be immediate).

I'm mulling solutions, along the lines of: http://stackoverflow.com/a/12613687/19212  – basically delaying the rendering of items that are off-screen.  My thinking is along the lines of, right [after insertion into the DOM](https://github.com/brianmhunt/knockout-fast-foreach/blob/master/index.js#L263):

``` javascript
// Per http://stackoverflow.com/a/7557433/19212
function isVisible(node) {
    var rect = el.getBoundingClientRect();
    return (
        rect.top >= 0 &&
        rect.left >= 0 &&
        rect.bottom <= window.innerHeight &&
        rect.right <= window.innerWidth
    );
}

childNode.forEach(function (node) {
   if isVisible(node) {
     // leave it alone
   } else {
     // insert a vertical placeholder the same height as the node,
     // then delay insertion of the node for a few milliseconds
     // so we don't get jank.
   }
})
```

Does that strategy sound like it'd hit the mark?

**Comment by krnlde** [comment]
The idea of deferring the actual draw is exactly what I need. But how would you tell the browser not to render the node? Deferring the `Node.appendChild()`? My guess is, if you defer node insertions to a timeframe of ~25ms each, it would kill the jank entirely - and as a result would build up the screen part by part. This, however, _could_ lead to unnecessary repaints of surrounding (depending) objects, but only if you are in a page flow and not positioning everything absolute, which I won't do in an SVG environment.

**Comment by brianmhunt** [comment]
Hmm.  A simpler option might be to batch childNodes by chunks of e.g. 50.

In that case we're adding an O(n) splicing operation to the node list, but the splicing is probably way less overhead than the also-linear-time DOM node insertion.

**Comment by krnlde** [comment]
:+1:  SGTM

**Comment by brianmhunt** [comment]
Thanks for reporting this @krnlde – It's an interesting use case.

I'll mull for a bit on the best way to do this.  @cervengoc Do you have any thoughts on this?

**Comment by krnlde** [comment]
My only objection: Rather than putting the chunks in a definite frame (like you said 50) I would listen for the heartbeat of the browser. `var start = Date.now()` before the operation and then lookup `if (Date.now() - start < 25) deferFollowingChunksToNextLoop()` in every iteration.
Browsers can profit from faster CPUs, while slower ones still don't jank.

**Comment by brianmhunt** [comment]
@krnlde I agree.

**Comment by cervengoc** [comment]
@brianmhunt I don't completely understand the use case and the problem on first read, and unfortunately I currently don't have time to get deeper into it.

One note: Chunking up the batch insertions shouldn't be so hard to implement. But I think for the sake of simplicity I would stay with a built-in constant chunk size instead of measuring the timing. The reason is that the code part which batches insertions is far before actually inserintg nodes. So let's say I would probably limit the batch insertions to 300, that would cover almost every usual scenarios I think. 

Hope I didn't misunderstood something or wrote something completely silly :)

**Comment by brianmhunt** [comment]
Thanks @cervengoc – appreciate the insight. :)

**Comment by krnlde** [comment]
The simplicity argument is signed. But what if you have a costly DOM insertion which itself needs 300ms? Then you would have 300x300ms of blocking iterations. In case of the time measurement the browser would insert one node per loop and, yes, still jank but not for a solid 90 seconds :-P

Don't get me wrong, I'm exaggerating the consequences of course. But the time measurement is just e few characters and 1 line more than to just check if the chunk size is reached. Plus, if your CPU is capable of rendering your full (say 10k) array in one 25ms time frame, you would prevent him from doing that by applying your chunk size regardlessly and thus slowing down the computation.

**Comment by cervengoc** [comment]
The cost of DOM insertion can widely vary, even data item by data item based on data structure, bindings, etc. It's really hard to measure it I think. 

In my personal opinion it should be very rare that someone has a template where insertion costs 300ms. Actually I wouldn't add the complexity to support it dynamically. If one has such a DOM structure, then he could set an option for chunk size explicitly.

**Comment by krnlde** [comment]
Not complicated at all, it's that easy:

``` javascript
var start = Date.now();
for (;;) {
  if (Date.now() - start < 25) appendNode();
  else return deferToNextEventLoop();
}
```

(pseudocode)

**Comment by cervengoc** [comment]
Okay, but with the batch insertion the inserted node is actually a documentFragment with several hundreds of list items. It's possible that I don't see something trivial though, but it seems not trivial for me to implement it.

If we didn't have the documentFragment utilization it would be more simple, but that optimization is key in much more often scenarios IMHO.

**Comment by brianmhunt** [comment]
@krnlde It's not quite that simple, as @cervengoc points out.

I think it's doable, but it'll take some rather cerebral thinking. :)

**Comment by krnlde** [comment]
Touché. For the ko world it might be true. That's why I'm requesting help here :)

Thank you guys for your time and your help so far.

**Comment by cervengoc** [comment]
By the way @brianmhunt how can it be that even when using documentFragment the browser still janks? Isn't inserting a documentFragment kind of an atomic thing?

**Comment by brianmhunt** [comment]
@cervengoc It totally is atomic, but as you can see from the "recalculating style" in the timeline capture @krnlde posted above, it can be a very long atomic thing. :)

**Comment by cervengoc** [comment]
Still I think the best way to give the ability to set a chunk size with an option. I think it' rare that a data collection is so heterogenous that one item takes 5ms and another takes 100.

The default could be to not use any chunking for keeping the performance, and developers could pass a desired value based on their knowledge about their data structure. 

What do you think?

**Comment by krnlde** [comment]
> it can be a very long atomic thing

lol. Which consequently can't be deferred to multiple event-loops. Also WebWorkers won't work here, because they [don't have access to the DOM](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers) ~~afaik~~.

**Comment by krnlde** [comment]
> The default could be to not use any chunking for keeping the performance, and developers could pass a desired value based on their knowledge about their data structure.

I can definitely commit to that.

**Comment by cervengoc** [comment]
Okay, so I'll start to think about it in the background but due to time reasons don't expect a solution thi year, at least not from me :)

**Comment by krnlde** [comment]
No probs. Contributions are always welcome, regardless what time :) Thank you!

**Comment by krnlde** [comment]
In the meantime, maybe I can have a look at that stuff. Can you provide some details how you would solve that around [the whole insertion](../blob/master/dist/knockout-fast-foreach.js#L309)? Sounds like [Promises](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Promise) are a good bet here. After a short look it appears to me that the only function depending on the node insertion is the `afterAdd` method which may or may not be provided through a `spec`. Am I right so far? That would mean we could insert the nodes completely asynchronously by first inserting the `ko.virtualElements.insertAfter(containerNode, frag, insertAfterNode);` ([#L309](https://github.com/brianmhunt/knockout-fast-foreach/blob/master/dist/knockout-fast-foreach.js#L309)) and _then_, step-by-step, insert the `frag.appendChild(nodeOrNodeArrayToInsert[i]);` on top of it asynchronously ([#L307](https://github.com/brianmhunt/knockout-fast-foreach/blob/master/dist/knockout-fast-foreach.js#L307)). After that we call the afterAdd method of the spec either through a Promise or directly via callbacks-style.

Also from what I can see the [afterQueueFlushed](/brianmhunt/knockout-fast-foreach/blob/master/dist/knockout-fast-foreach.js#L238) method is called after each cycle and not at the end of the whole queue, which would inform listeners over a partially ready list and not over a whole list - maybe that was your guys intention so you can react on the nodes immediatelly. But then not only the name "queue flushed" is wrong but also the latter usecase is missing.

**Comment by cervengoc** [comment]
@krnlde Thank you for your interest and efforts. However, before looking into this I would really like to see #34 merged, because I've made major changes in internal logic which surely affects a proposed implementation of this as well. @brianmhunt will probably merge it soon :)

**Comment by krnlde** [comment]
I stay tuned and subscribe to that issue :)

**Comment by krnlde** [comment]
This sounds awfully familiar: https://github.com/knockout/knockout/issues/1965

**Comment by vamp** [comment]
@krnlde, do not use large lists (the better solution would be - [clusterize.js](https://github.com/NeXTs/Clusterize.js) or [clusterize binding for knockout](https://cloudup.com/cBsdZgzd9iB))

**Comment by krnlde** [comment]
Hey @vamp! Thank you for your answer, but why shouldn't I use large lists and let _"the"_ library decide how to handle it. As the title suggests, a chunked, asynchronous approach could help.
Also clusterize is no solution when drawing a lot of dots in one sivible SVG.

**Comment by vamp** [comment]
@krnlde, sample for you: https://jsfiddle.net/312j519d/

**Comment by krnlde** [comment]
Draw these as dots in an SVG, without scrolling  and we have a deal...

**Comment by krnlde** [comment]
Since #34 (and #33 respectively) are closed, can we get back to this issue @brianmhunt ? I linked a knockout issue a few days back which describes a similar behaviour. Any insights?

**Comment by brianmhunt** [comment]
I think if we were to tackle this, the best option might be to mimic clusterize – at least the part that renders and binds the elements only when they are in the viewport.

That would address the rendering-side, though it means that the item-bindings will be lazy. This breaks the assumption that everything in the list will have bindings applied immediately, but for what'll be a massive performance boost, fair enough.

Mass-removes will still be a problem, as in #37 ...

**Comment by krnlde** [comment]
Clusterize, at least to me, is only a workaround for a yet to be solved problem. It doesn't nip things in the bud. We rather need a possibility to defer the browser painting over a few loops, so the browser never janks. That would resolve every problem - btw. including that one clusterize tackles. The only solution we need to find is to tell the browser to just draw some nodes and to join the asynchronous events to one synchronized one, the `afterQueueFlush`, but `Promise` is a good bet here.

**Comment by brianmhunt** [comment]
Thanks @krnlde – Solid input.

We have access to a couple async facilities, namely `Promise`, the `ko.task` microscheduler, and the `window.animateFrame` (which has some side benefits).  We will have to experiment to see what works best.

**Comment by krnlde** [comment]
I need to get deeper into the knockout and FFE code before I can provide some experimental code. :-/ Right now the mechanics of knockout is a little hard to grasp, especially all the bindingHandler parameters. Too confusing.

**Comment by cervengoc** [comment]
Hello guys and first of all Happy new year :)

I had some time to work on an experimental implementation. Please check it out at my fork: https://github.com/cervengoc/knockout-fast-foreach

I'm chunking the node insertions at the "last" moment to keep the internal state flow untouched. With this delayed insertion I also had to add a check when removing nodes to ensure that the node is attached to the DOM at that moment.

I modified the `afterAdd` logic so that it runs for every chunk. Using this, I've prepared only one unit test so far because this behavior is somewhat hard to unit test.

Looking forward to hear the test results and your suggestions.

**Comment by brianmhunt** [comment]
Thanks @krnlde – yes, the mechanics are not clear.  Hopefully it's something we'll improve in the not-too-distant future.

@cervengoc Your approach is what I was thinking, too.

The test you have is good.  The two other scenarios we probably ought to test for are:
1. Adding has started, and new items are added before the all existing items are inserted into the DOM; and
2. Adding has started, but the items being added are deleted before being inserted into the DOM.

It's not-trivial to test this stuff.

Feel free to start a PR anytime (or just create a branch off this repo, which you should now be able to do as a collaborator).

Cheers

**Comment by cervengoc** [comment]
Yes, I was thinking about your second test idea, but need some time to figure out how to cover these in the testing environment.

In the meantime @krnlde could you please test my version? Now it has a chunk size of 100 built in, in the final version it will be configurable.

**Comment by krnlde** [comment]
At a first glance your implementation looks exactly like what I need. I'll drop it in now and test with my SVG what happens.

**Comment by krnlde** [comment]
Superfast! Just for the record: We are now drawing 12k dots in an SVG map **without even a single jank** in Firefox and Chrome.

In IE11 and Edge you can watch the browser drawing, but at least you can scroll and the browser responds during that process (with a little jank).

![image](http://i.imgur.com/jctq49A.gif)
Wheeeee...

**Comment by cervengoc** [comment]
I'm glad to here the news :) so now I'll continue the tests and hope
nothing has broken.

**Comment by krnlde** [comment]
I've created a profile for Edge during the drawing of the dots, this is what it looks like:

![profile](https://cloud.githubusercontent.com/assets/1087002/12149043/f1d3a0ca-b4a1-11e5-93cc-36535024efc4.PNG)
(Sorry for the german labeling)

The anonymous function callback I focused is exactly [this line](https://github.com/cervengoc/knockout-fast-foreach/blob/master/index.js#L169):
`FastForEach.animateFrame.call(window, function () { self.processQueue(); });`

I'm not sure why Edge still janks. Maybe it is simply too slow to even draw 100 dots in a loop. But it's funny to see it being built up.

**Comment by cervengoc** [comment]
The `processQueue` method shouldn't be called so many times. Actually it should be called only once if you add the items all together (like using the `valueWillMutate` + `ko.utils.arrayPushAll` technique). Are you sure it's this anonymous method and not the other one which calls the rendering of the next chunk?

**Comment by krnlde** [comment]
I'm sure the first anonymous function is `FastForEach.animateFrame.call(window, function () { self.processQueue(); });`. I'm not 100% sure about following functions, but I'll test that again as soon as I have my environment up and running again.

**Comment by cervengoc** [comment]
Yes, the first should be that, and the subsequents should be the chunked
rendering.
I'll make the chunk size configurable later today, so you can test IE with
a smaller number if you wish.

**Comment by krnlde** [comment]
The problem is, that IE seems to be hanging while all those chunks are processed, it doesn't draw anything in that time, not even CSS animations.

**Comment by cervengoc** [comment]
Ok, then unfortunately my current knowledge is not enough even for any idea
about it. @brianmhunt do you have any thought about IE behaviour?

**Comment by krnlde** [comment]
I'll test it again with more profiling insights and share the results with you guys.

**Comment by cervengoc** [comment]
Ok, thank you for your efforts.

2016-01-08 15:10 GMT+01:00 Kai Dorschner notifications@github.com:

**Comment by brianmhunt** [comment]
Hmm.  I wonder if it's an issue with how IE/Edge implements `requestAnimationFrame`.

What is the value of `ko.bindingHandlers.fastForEach.FastForEach.animateFrame`?

What happens when `ko.bindingHandlers.fastForEach.FastForEach.animateFrame = window.setTimeout(cb, 1000 / 60);`?

**Comment by krnlde** [comment]
I hate IE so much. ;)

**Comment by brianmhunt** [comment]
Yeah, if it works in Safari + Chrome + Firefox, then I'd be inclined to report the problem upstream. ... if there was an upstream to whom to report it. :)

**Comment by cervengoc** [comment]
I think nowadays one can almost always afford such minor issues of IE. It's important to think about kind of a "price/performance" of a feature/fix/workaround. In one word, I wouldn't care about it too much if it works in other major browsers like this.

**Comment by krnlde** [comment]
I'll have it tested in Safari as well. I think in an hour I am good to go.

**Comment by krnlde** [comment]
![image](https://cloud.githubusercontent.com/assets/1087002/12202811/fcc4c1e0-b62b-11e5-8aa8-d962e85a340a.png)

I tested it again with around 2k dots. Right now I can only provide a Timeline from Chrome but here happens the same, the processQueue is called very often.

Here is a detail of a single rAF call:

![image](https://cloud.githubusercontent.com/assets/1087002/12202907/797cf522-b62c-11e5-99ac-032a0ad9bf73.png)

What makes me wonder is that the frame size is so long (2115ms). I'm sure it is not the time the browser janks, it is just the time it needed to compute the array (while being reactive at that time, thanks to rAF). The flame chart breaks down to really nice and short atomic instructions - so at least this is no concern.

**Comment by brianmhunt** [comment]
Weird.  The rAF should be at-most 1/60th of a second.  The chain should look like this:

```
- processQueue
  - animateFrame
    - ...
    - insertAllAfterByChunks
      - animateFrame
        - insertAllAfterByChunks
          - animateFrame
```

In other words, there should be one, and only one, top-level processQueue for each `arrayChange` 
event.

Perhaps this is what is happening, but the presentation (for the rAF) is "up-ended" so the animateFrame appears to be the caller – when in reality it's the processQueue.  In other words, visually it is many calls to processQueue when in reality it is one call spread out over many frames.  

What's the output when you do a `console.log` in processQueue? (i.e. to count the number of times it's called)

**Comment by krnlde** [comment]
When I console log _inside_ processQueue it shows the log exactly as often as in the flame chart. oO

OK, but this could be the nature of my app. Since I have other things going on and I also have nested foreach's

**Comment by krnlde** [comment]
Alright, just clearing things up: When I also put a `console.log` inside `FastForEach.prototype.insertAllAfterByChunks` I see exactly what should happen. As said the amount of processQueues is in because of my special setup. The pure dots are computed in exactly one processQueue with lots of insertAllAfterByChunks. It is just funny that I don't find that function in the flame chart. I only see the `FastForEach.prototype.added` which is as big as the processQueue.

**Comment by krnlde** [comment]
Found it, is on the end of the `FastForEach.prototype.added` function:
![image](https://cloud.githubusercontent.com/assets/1087002/12203679/8d450460-b630-11e5-9caf-5f092b73c7c3.png)

**Comment by krnlde** [comment]
A test with ~12k dots scales in the same manner in Chrome. No jank, just a little more time to draw the dots. We (I) should definitely reduce the chunk size, because it only draws 1-2 frames each sec.

Edge has bigger problems hovever. These frames I was mentioning earlier seem to block the browser entirely - for solid 5-6 secs on ~12k dots. :-/

![image](https://cloud.githubusercontent.com/assets/1087002/12204528/fc24bed0-b634-11e5-8190-bd7df1d821bb.png)

But as @cervengoc said, it's just IE/Edge... I think we made awesome progress on fast-foreach and I definitely appreciate the work you guys have done (fortunately not only for me :+1: )! Thank you!

**Comment by cervengoc** [comment]
Hi, I've updated my fork to allow configuring the chunk size. I've also fixed a critical issue as the DOM node chunks got inserted in reverse order.

However, I'm still not sure if we should allow the chunk size configuration at a per-binding level, or it's enough to test and find an optimal value for "everyone". I think that this option is too "technical" to include as a binding option, but will see.

For now you can globally set `FastForEach.DOM_INSERTION_CHUNKSIZE`, which has a default value of `50`.

Also, I've found some cases where the internal logic is affected and I have to think it over a few more times, so this is still far from being stable and ready to be merged.

**Comment by krnlde** [comment]
The biggest concern would be a modification on the array while `insertAllAfterByChunks`. You sould lock the array for that time and apply the modification as soon as the insertion is done.

**Comment by cervengoc** [comment]
Yes, something like that. Removes are not problem, but any insert on an index which has a pending DOM insertion could break it all at first glance.
The safest way would be to delay the `processQueue` call until the insertions complete. This could be easily implemented with an internal state flag. Then the `processQueue` would delay itself until it has a "free" frame. This way we could overcome every "conflict" scenarios. @brianmhunt what do you think?

**Comment by krnlde** [comment]
Hey guys! It's been a while we talked so I just wanted to bump this topic since all our precious work we've done gets lost if we do not merge it back to the master.

How can I help?

**Comment by cervengoc** [comment]
@krnlde Unfortunately I'm completely drowning in my current work, and don't have time for it now. Possible in a few weeks it will get lighter.

As I wrote earlier, the delaying of `processQueue` would be a good solution I think to handle concurrent DOM inserts and array modifications.

Also, it would be needed to cover some cases with new tests, and I have a feeling that some of them will not be trivial.

**Comment by krnlde** [comment]
I could write some tests I guess. Concurrent modifications are indeed a problem. But that should be testable, too.

I feel not comfortable enough to edit the plugin  though, so unfortunately no library specific help for `processQueue` from my side :-/

**Comment by cervengoc** [comment]
I can make that modification, the coding itself should not be a big deal, but I have no time to design and write tests for it. So if you can help with that, I would appreciate it. I will try to make these changes in the next days.

**Comment by cervengoc** [comment]
@krnlde @brianmhunt Sorry for the huge delay, I finally had some time to implement the aforementioned synchronization between queue processing and DOM manipulations (insertions for now).You can find it in my fork's latest code.

The implementation is simply based on a flag which is set until chunked DOM insertions finish. If this flag is set, the `processQueue` method simply delays itself for the next frame, and so on.

Note that I unfortunately didn't have time for preparing tests for all of this stuff, would be great if someone could join the party :)

**Comment by krnlde** [comment]
Hey @cervengoc thank you for your implementation. Where should I post issues to this? Here or in your repo?

**Comment by cervengoc** [comment]
I think we should move this topic there for clarity.

**Comment by cervengoc** [comment]
@brianmhunt @krnlde Sorry guys, I messed up some things. I completely reverted the previous implementation and found another one, which you can find in my fork. I had to recreate the complete fork because I'm such a noob with git, I am suffering a lot with more complex tasks.

The problem with the previous implementation was that it partitioned only the DOM additions or large node sets. However, in many cases the whole processing can be a lot of work (eg. applying bindings to all children when bindings are complex, for example including nested `fastForEach` bindings, etc.).

So the new implementation partitions the whole processing. I'll write some comments on the commit.

**Comment by krnlde** [comment]
You were faster than me :) I'll hold back the bug report and test the new implementation. Thank you again!

**Comment by krnlde** [comment]
@cervengoc unfortunately I can't create issues in your fork so this'll have to do: I created an isolated scenario with your new fastForEach implementation here: https://jsfiddle.net/krnlde/huhz0p9x/

Each block contains 3200 circles.

You can see in line 485 that I am timing the actual draw by starting the timer and stopping it the (quasi) next event loop.

Now open the console in Chrome. When you hover over the gray blocks you'll see a lot of white circles appearing and as you mouseout they'll disappear. The process is still synchronous and blocks the browser for around 250ms avg on my PC. Plus on my real world example the second mouseover on a block (after a fresh page reload) is incredibly slow to draw (around the factor "times 10"). After that slow draw, everything goes back to normal. Could this have anything to do with the nested FFE?

Hope I could help a bit

**Comment by cervengoc** [comment]
@krnlde Thanks for the feedback. I'm sorry I forgot to mention that the new implementation is an opt-in setting, which means that by default it doesn't use partitioning.

You have to set the `partition` option at binding, like 

```
data-bind="fastForEach: { data: items, partition: 50 }"
```

Also, note that measuring time can take some resources itself, so please at first step, test without it.

**Comment by krnlde** [comment]
This works really good! Thank you for this. Only issue right now is in a nested fastForEach the partition would run in parallel. For example if you have 10 rows and 20 cols and you iterate over them with fastForEach like so:

``` html
<!-- ko fastForEach: {data: rows, partition: 5} -->
<div class="row">
  <!-- ko fastForEach: {data: cols, partition: 5} -->
  <div class="col">
    ...
  </div>
  <!-- /ko -->
</div>
<!-- /ko -->
```

in the first iteration it'll draw 5 rows and 0 cols. In the second iteration it'll draw another 5 rows but now it'll draw 5 cols in each of the previously drawn rows, so it is `5 * 5 = 25` rows in the second iteration.

What I would expect that every partition is globally queued and processed. Where should I put this issue @cervengoc? Could you open the "Issues" Tab in your forked FFE so we can write there? Also I need some advice on how to write and execute the tests.

**Comment by krnlde** [comment]
Bump

**Comment by cervengoc** [comment]
Sorry guys I'm having a holiday,  I'll be back next week.
2016. jún. 17. de. 11:24 ezt írta ("Kai Dorschner" <notifications@github.com

**Comment by cervengoc** [comment]
@krnlde I'll try to enable Issues in my fork.

However, this nested behavior is IMO the expected. I think it's a very rare real-world need to nest two fast-foreach instances each having a defined partition size, because in any case which I can think of you have one "dimension" which is "heavy" and others not.

Also, in pure technical speaking it would add a big complexity to introduce some kind of global queue handling and IMO it's far not worth it. My current implementation is I think simple enough to find a way to the main fork.

**Comment by cervengoc** [comment]
Are here any news? I had to restructure my fork, but the implementation is still there in its own branch.

**Comment by krnlde** [comment]
From my side everything is fine. I wrote the things that bothered me, but apart from that there's only the tests that need to be written. I was asking how to do it. Could you clarify?

**Comment by cervengoc** [comment]
Ok. This is not easy to test in my opinion. One option would be to fake the `animateFrame` function of fastForEach so it would provide some kind of callback where we can check after each cycle if the partitioning really happens, eg. the appropriate number of child elements are there after each cycle.

What I did for the sake of quick check is duplicated all the tests with a non-zero partition size to see if it breaks anything. But I don't like this testing technique, so we'll have to prepare some _normal_ tests.

For test running we use `karma`, and for the unit tests and assertions `mocha` and `chai` currently. All the tests are in one big `js` file now in the folder `spec`. If you check the existing tests, you can have an idea what's happening.

But anyway, when I have some time, I can prepare those tests too. Just don't know when yet.
